### PR TITLE
TSDB: compactor's syncer metrics

### PR DESCRIPTION
**Pull request by pstibrany**
This PR exposes metrics from TSDB's compactor "syncer" component. In Cortex, we instantiate multiple syncers, new one for each sync operation. Because of this, approach used to gather metrics is different from previous TSDB metrics work.

Here we register fixed set of metrics that are exported, and we update them when syncer has done its job, via `defer c.syncerMetrics.gatherThanosSyncerMetrics(reg)` in `compactUser` method.

Since tsdb.Compactor is reused between `compactUser` calls, we don't need to handle those metrics (they are expored with their original `prometheus_tsdb_compactions` prefix)

**Checklist**
- [x] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/compactor/syncer_metrics_test.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
+package compactor
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/compactor/syncer_metrics_test.go]: Good idea. I just felt that test is too big to put it into compactor, but you're right that we can easily extract syncer_metrics.go. Done.

File: pkg/compactor/compactor.go [hunk 2]
```diff
@@ -261,3 +268,128 @@ func (c *Compactor) discoverUsers(ctx context.Context) ([]string, error) {
...
+
+	m.syncMetas.Add(mfm.SumCounters("thanos_compact_sync_meta_total"))
+	m.syncMetaFailures.Add(mfm.SumCounters("thanos_compact_sync_meta_failures_total"))
+	m.syncMetaDuration.Add(mfm.SumHistograms("thanos_compact_sync_meta_duration_seconds"))
+	m.garbageCollectedBlocks.Add(mfm.SumCounters("thanos_compact_garbage_collected_blocks_total"))
+	m.garbageCollections.Add(mfm.SumCounters("thanos_compact_garbage_collection_total"))
+	m.garbageCollectionFailures.Add(mfm.SumCounters("thanos_compact_garbage_collection_failures_total"))
+	m.garbageCollectionDuration.Add(mfm.SumHistograms("thanos_compact_garbage_collection_duration_seconds"))
+
+	// There have "group" label, but we sum them all together.
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/compactor/compactor.go]: I meant "These metrics" :)

File: pkg/util/metrics_helper.go [hunk 3]
```diff
@@ -4,77 +4,147 @@ import (
...
+		}
+	}
+	return data
 }
 
 func (d MetricFamiliesPerUser) SendSumOfCounters(out chan<- prometheus.Metric, desc *prometheus.Desc, counter string) {
 	result := float64(0)
-	for _, perMetric := range d {
-		result += sum(perMetric[counter], counterValue)
+	for _, perUser := range d {
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/util/metrics_helper.go]: Now it's `userMetrics` everywhere.

File: pkg/compactor/syncer_metrics.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	garbageCollectionDuration *util.HistogramDataCollector // was prometheus.Histogram before
+	compactions               prometheus.Counter
+	compactionRunsStarted     prometheus.Counter
+	compactionRunsCompleted   prometheus.Counter
+	compactionFailures        prometheus.Counter
+	verticalCompactions       prometheus.Counter
+}
+
+// Copied (and modified with Cortex prefix) from Thanos, pkg/compact/compact.go
+// We also ignore "group" label, since we only use single group.
```
Comment (jtlisi) [inline pkg/compactor/syncer_metrics.go]:
```suggestion
// We also ignore "group" label, since we only use a single group.
```

Comment (pstibrany) [reply pkg/compactor/syncer_metrics.go]: Accepted your suggestion, which made you a co-author ;-)

**Pull request merged**